google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITQueryCountTest.java

```diff
@@ -0,0 +1,336 @@
+      DocumentReference doc = collection.document();
+      writeBatch.create(doc, singletonMap(key, value));
+    }
+    writeBatch.commit().get();
+  }
+
+  private void createDocumentInCollection(WriteBatch writeBatch, CollectionReference collection)
+      throws ExecutionException, InterruptedException {
+    createDocumentInCollection(writeBatch, collection, "age", 42);
+  }
+
+  private void createDocumentInCollection(
+      WriteBatch writeBatch, CollectionReference collection, String key, int value)
+      throws ExecutionException, InterruptedException {
+    DocumentReference doc = collection.document();
+    writeBatch.create(doc, singletonMap(key, value));
+  }
+
+  private CollectionGroup createCollectionGroupWithDocuments(int numDocumentsToCreate)
+      throws ExecutionException, InterruptedException {
+    String collectionId = autoId();
+
+    // Create some collections to participate in the group.
+    ArrayList<CollectionReference> collections = new ArrayList<>();
+    for (int i = 0; i <= numDocumentsToCreate / 3; i++) {
+      collections.add(createEmptyCollection().document().collection(collectionId));
+    }
+
+    // Populate the collections with documents.
+    WriteBatch writeBatch = firestore.batch();
+    Iterator<CollectionReference> collectionIterator = loopInfinitely(collections);
+    for (int i = 0; i < numDocumentsToCreate; i++) {
+      createDocumentInCollection(writeBatch, collectionIterator.next());
+    }
+
+    writeBatch.commit().get();
+
+    return firestore.collectionGroup(collectionId);
+  }
+
+  /** Converts a {@link Timestamp} to the equivalent number of milliseconds. */
+  private static long msFromTimestamp(Timestamp timestamp) {
+    return (timestamp.getSeconds() * 1_000) + (timestamp.getNanos() / 1_000_000);
+  }
+
+  /**
+   * Creates and returns an infinite iterator that iterates over the given collection in the same
+   * order as its {@link Collection#iterator} method does, looping around back to the beginning when
+   * it reaches the end.
+   */
+  private static <T> Iterator<T> loopInfinitely(Collection<T> collection) {
+    return new Iterator<T>() {
+      private Iterator<T> it = collection.iterator();
+
+      @Override
+      public boolean hasNext() {
+        // NOTE: it.hasNext() will ALWAYS return true, except in the case that the given collection
+        // is empty.
+        return it.hasNext();
+      }
+
+      @Override
+      public T next() {
+        T element = it.next();
+        // If we've reached the end of the iterator, create a new one.
+        if (!it.hasNext()) {
+          it = collection.iterator();
+        }
+        return element;
+      }
+    };
+  }
+
+  /**
+   * Blocks the calling thread until the given future completes. Note that this method does not
+   * check the success or failure of the future; it returns regardless of its success or failure.
+   */
+  private static void await(ApiFuture<?> future) throws InterruptedException {
```

Comment:
Can we use `future.get()`?

Reply:
future.get() will throw the exception if the task fails, which isn't the desired behavior for the singular use of this function.

Reply:
Will this do:

```
try {future.get()} finally {assertTrue(future.done());}
```

?